mp/src/game/server/momentum/mom_triggers.cpp

```diff
@@ -523,18 +717,87 @@ void CTriggerTrickZone::OnEndTouch(CBaseEntity *pOther)
     }
 }
 
+//----------------------------------------------------------------------------------------------
+
+bool TeleportEntity(CBaseEntity* pTeleportTo, CBaseEntity* pEntToTeleport, int iMode = 0, Vector vecVelocityScaler = Vector(1.0f, 1.0f, 1.0f), bool bResetAngles = true)
+{
+    if (!(pTeleportTo && pEntToTeleport))
+        return false;
+
+    pEntToTeleport->SetGroundEntity(nullptr);
+
+    QAngle* pAngles = const_cast<QAngle*>(&pTeleportTo->GetAbsAngles());
+    Vector* pVelocity = const_cast<Vector*>(&pEntToTeleport->GetAbsVelocity());
+
+    matrix3x4_t matMyModelToWorld;
+    VMatrix matMyInverse, matRemoteTransform;
+    QAngle angVelocityAngle;
+
+    VectorMultiply(*pVelocity, vecVelocityScaler, *pVelocity);
+
+    switch (iMode)
+    {
+    // Redundant due to the velocity scaler but needed to maintain compatibility with existing maps that use momentum teleports
+    case TELEPORT_RESET:
+        pVelocity = const_cast<Vector*>(&vec3_origin);
+        break;
+
+    case TELEPORT_KEEP_NEGATIVE_Z:
+        pVelocity->x = 0;
+        pVelocity->y = 0;
+
+        if (pVelocity->z > 0.0f)
+            pVelocity->z = 0;
+
+        break;
+
+    case TELEPORT_SNAP_TO_DESTINATION:
+        // Build a transformation from the teleportee's velocity vector
+        VectorAngles(pEntToTeleport->GetAbsVelocity(), angVelocityAngle);
+        AngleMatrix(angVelocityAngle, matMyModelToWorld);
+        MatrixInverseGeneral(matMyModelToWorld, matMyInverse);
+        matRemoteTransform = pTeleportTo->EntityToWorldTransform();
+
+        // Reorient the velocity
+        *pVelocity = matMyInverse.ApplyRotation(*pVelocity);
+        *pVelocity = matRemoteTransform.ApplyRotation(*pVelocity);
+        break;
+
+    default:
+        // vphysics objects get stopped even when no velocity is sent, so just send the current velocity
+        if (pEntToTeleport->GetMoveType() != MOVETYPE_VPHYSICS)
+            pVelocity = nullptr;
+    };
+
+    if (!bResetAngles)
+        pAngles = nullptr;
+
+    pEntToTeleport->Teleport(&pTeleportTo->GetAbsOrigin(), pAngles, pVelocity);
+
+    return true;
+}
+
 //----------- CTriggerTeleport -----------------------------------------------------------------
 LINK_ENTITY_TO_CLASS(trigger_momentum_teleport, CTriggerMomentumTeleport);
 
 BEGIN_DATADESC(CTriggerMomentumTeleport)
-    DEFINE_KEYFIELD(m_bResetVelocity, FIELD_BOOLEAN, "stop"),
+    DEFINE_KEYFIELD(m_iMode, FIELD_INTEGER, "stop"),
```

Comment:
Should this change its keyvalue to be `"mode"`?

Reply:
I left it as is because there might already be some maps that use this trigger